### PR TITLE
Shadow the polymorphic comparison in the middle-end

### PR DESCRIPTION
**Pull request by xclerc**
Almost everything is in the title...

This PR introduces a module, namely `Int_replace_polymorphic_compare`,
that defines the `compare` function and the comparison operators to work
on `int` values. The module is then opened at the top of every module in
the middle-end. This guarantees that the polymorphic comparison cannot
be used by mistake (this PR being a follow-up of https://github.com/ocaml/ocaml/pull/1810).

(There were suspicious comparisons in `Flambda_utils.sameswitch`, where `=`
was used to compare values of type `Numbers.Int.Set.t`.)

### CODE REVIEW

**Comment by bluddy** [comment]
Polymorhphic comparison strikes again!

**Comment by xclerc** [comment]
@chambart I am looking for a reviewer; would you be interested?

File: middle_end/flambda_utils.ml
Comment (kit-ty-kate) [inline middle_end/flambda_utils.ml]: What is the type of `p1` and `p2` here ?

Comment (xclerc) [reply middle_end/flambda_utils.ml]: The type is `Lambda.primitive`, which has more than 100 constructors.
I reckon something should be done here; either write an equality function
for `Lambda.primitive` (and the types it references), or at the very least
ensure that the type does not change over time. However, one can
genuinely be worried that one day a float value might pop up in the
definition of `Lambda.primitive` (or one of its numerous referenced types).

Comment (xclerc) [reply middle_end/flambda_utils.ml]: (For the record, and the impatient reader, this is the only place
where I kept the polymorphic comparison.)

Comment (yallop) [reply middle_end/flambda_utils.ml]: Perhaps it'd be worth adding a `Lambda.primitive_equal` function (initially an alias for `Stdlib.(=)`) to keep the definition of equality close to the definition of the type.

Comment (xclerc) [reply middle_end/flambda_utils.ml]: @yallop Sure; if memory serves I put all introduced equality functions
in the same modules as the type definitions.

Comment (kit-ty-kate) [reply middle_end/flambda_utils.ml]: Maybe it might be worth adding a comment to `Lambda.equal_primitive` to know why it uses polymorphic equality.

Comment (xclerc) [reply middle_end/flambda_utils.ml]: Indeed; should I say that I am too lazy to write the equality
function because there are too many constructors and nested
types involved?

Comment (kit-ty-kate) [reply middle_end/flambda_utils.ml]: Sure, it's a totally valid comment ^^

**Comment by dra27** [comment]
There are some lines over 80 characters. #1148 will be merged soon, which will turn that into a CI error:

```
./bytecomp/lambda.mli:217.81: [long-line] line is over 80 columns
./middle_end/flambda_utils.ml:81.81: [long-line] line is over 80 columns
./middle_end/inlining_cost.ml:484.81: [long-line] line is over 80 columns
./middle_end/inlining_transforms.ml:113.81: [long-line] line is over 80 columns
./middle_end/simplify_boxed_integer_ops.ml:73.81: [long-line] line is over 80 columns
./middle_end/simplify_boxed_integer_ops.ml:74.81: [long-line] line is over 80 columns
./middle_end/simplify_boxed_integer_ops.ml:88.81: [long-line] line is over 80 columns
./middle_end/simplify_primitives.ml:227.81: [long-line] line is over 80 columns
./middle_end/simplify_primitives.ml:228.81: [long-line] line is over 80 columns
./middle_end/simplify_primitives.ml:229.81: [long-line] line is over 80 columns
```

**Comment by xclerc** [comment]
@dra27 the reported issues should be fixed by `6595154`.

By the way, I am still looking for a reviewer; @chambart maybe?

**Comment by dra27** [comment]
@xclerc - it does indeed, thanks. The failure on Travis is a subtle bug in the CI script, which I just fixed in https://github.com/ocaml/ocaml/commit/f395dfc96b85926afe2235c6f0de498d0f0d2386, so the failure can either be ignored or should disappear on any future push to this branch.

File: middle_end/flambda.ml
Comment (chambart) [inline middle_end/flambda.ml]: This may go into Misc.Option

**Review by chambart — APPROVED** [review]
This looks good to me, and fixes some problems.

**Comment by chambart** [comment]
I'm happy to merge when the check-typo test will pass

**Comment by dra27** [comment]
check-typo failed originally because of a bug in the script (it read the files from an older commit, so the failure was already fixed on trunk). Now it's being weird again as I restart it!

However: this needs a `Changes` entry and pushing that should hopefully kick the check-typo job. If it doesn't, I'll take it from here 🙂

**Comment by xclerc** [comment]
Everything is green - ready to merge!

**Comment by damiendoligez** [comment]
@xclerc alas GitHub says you need to rebase and there will be conflicts.

**Comment by xclerc** [comment]
@damiendoligez I am slightly confused: the github interface says
"This branch has no conflicts with the base branch".

**Comment by damiendoligez** [comment]
Well I'm confused too: what GitHub tells *me* is:
```
This branch cannot be rebased due to conflicts
```

screen shot 2018-07-09 at 16 21 08

**Comment by dra27** [comment]
@damiendoligez - it may be that it can be merged, but it can’t be rebased (that would make sense, as trunk has been merged into this branch)

**Comment by xclerc** [comment]
@damiendoligez So, can this PR be merged on your end?

**Comment by damiendoligez** [comment]
@dra27 is right, I can squash-and-merge or do a merge commit, but I cannot rebase-and-merge, most likely because you merged trunk instead of the usual (rebasing the PR on top of trunk).

So, @xclerc do you think the commit history is useful (in which case I should merge) or should I just squash everything into one commit?

**Comment by xclerc** [comment]
@damiendoligez I don't think the history is useful here; you can squash
everything into one commit. Thanks!

**Comment by damiendoligez** [comment]
Thanks!

**Comment by gasche** [comment]
I just saw a build failure on the "parallel build" CI machine whose log (below) suggests that it *might* be related to a missing `make depend` step after this PR is merged.

```
File ".../workspace/parallel-build/flambda/false/label/ocaml-linux-64/_none_", line 1:
Error (warning 58): no cmx file was found in path for module Int_replace_polymorphic_compare, and its interface was not compiled with -opaque
Makefile:1293: recipe for target 'middle_end/debuginfo.cmx' failed
```

**Pull request merged**
